### PR TITLE
Optimize DSA Gather4 and score kernels; document sparse attention

### PR DESCRIPTION
**Pull request by jiayus-nvidia**
## Summary

Keep SM100 DSA Indexer Backward on TMA Gather4 for local top-k IDs whenever the installed DSL supports Gather4. Local IDs at top-k 512/1024/2048 previously selected the manual loader; the existing in-register bounds checks and batch-offset conversion now feed the optimized gather path.

Recover score-kernel performance lost with CuTeDSL 4.6.2:

- Explicitly mark the shared dense score kernel's warpgroup ID as warp-uniform. This avoids redundant collective control flow in indexer forward, dense indexer score recompute, fused indexer forward Top-K, and dense attention score recompute.
- Compile **n128 sparse attention score specializations with O2** to avoid register spilling observed at O3. Select by `n_block_size`, without pinning a DSL version, head count, head dimension, or Top-K. Keep n64 at its default optimization level: O2 can regress that specialization. This covers both non-compact n128 and the compact H128/D512 n128 configuration.
- Add 20 numerical reference cases spanning n64/n128, K splitting, H32/H64/H128, D128/D512/D576, Top-K 128/512/1024, local/global IDs, optional lengths, and partial final tiles.

Expand DSA documentation on ID representations, compact sparse indices, memory policies, and backend selection. Add the Flex Attention mask-plan design and three GB300 reference benchmark figures with source revision and pre-migration provenance, linked from the documentation indexes.

## Why

Local ID encoding should not disable Gather4 or require an extra conversion launch for Indexer Backward. Independently, changing the DSL compiler exposed excess collective control flow in the dense kernel and register spilling in n128 sparse attention score kernels. The uniformity annotation conveys an existing invariant (`warpgroup_size=128`, warp size 32); the n128 compile option removes the observed spills without applying O2 to n64 configurations that can slow down.

The exact compiler pass responsible has not been bisected. O2 is a measured specialization choice, not a claim that lowering optimization improves every kernel.

## API and compatibility

Public signatures and local/global ID semantics are unchanged. The manual backward loader remains available when Gather4 is unavailable. No DSL dependency-floor change or new version-specific API dependency (Python Rule 7). Compilation options depend on the existing n-block specialization already present in the cache key.

Related to #775 for Flex Attention migration. Affected areas: FE OSS/CuTeDSL, DSA regression coverage, and documentation.

## Validation

B200 (SM100), CuTeDSL 4.6.2, CUDA runtime 13.2, cuDNN 9.22, PyTorch `2.12.0a0+5aff3928d8.nv26.05`. Verified the actual `cudnn` import resolves to the PR worktree before testing and benchmarking.

From the PR worktree's `test/python`:

```bash
PYTHONPATH=/code/github/cudnn-frontend-pr957/python \
/tmp/pr957-tests/bin/python -m pytest -q -m 'L0 or L1' \
  fe_api/dsa/test_DSA_sparse_score_recompute.py \
  fe_api/dsa/test_DSA_dense_score_recompute.py \
  fe_api/dsa/test_DSA_indexer_forward.py \
  fe_api/dsa/test_DSA_indexer_forward_compressed.py
```

**65 passed, 6 skipped**, including all 20 added reference cases. The new reference cases validate numerical behavior under the selected compile options; the performance evidence is the A/B measurement below, not a timing assertion in pytest.

Earlier Gather4 validation in this PR: 11 tests passed covering local/global IDs, gradients, and the manual-loader fallback. Restoring the old dispatch made all three `local_ids_use_tma_gather` tests fail at the expected assertion; restoring the fix made them pass. That backward-only selection was not rerun for the score-kernel addition.

### Score-kernel performance

Compare PR head `65dba0cf8` with this addition on the same B200 and DSL 4.6.2. BF16; B=2, Sq=4096, Sk=131072, H=64, ratio=4, causal offsets `[0, 520192]`; D=128 for indexer and D=512 for attention; Top-K=512. Public wrappers, default tiles/stages, first JIT excluded, 10 warmups, median of 30 independent CUDA-event samples. Units: milliseconds.

| API / variant | Before | After |
|---|---:|---:|
| sparse_attn_score_recompute / None (n128) | 0.6817 | 0.5247 |
| sparse_attn_score_recompute / provided (n64 control) | 0.5183 | 0.5179 |
| dense_indexer_score_recompute | 9.3575 | 8.3006 |
| indexer_forward_top_k | 11.2697 | 10.5327 |
| dense_attn_score_recompute | 29.4526 | 28.2057 |

A separate O2/O3 sweep used B=2, Sq=512, Sk=16384, BF16, 15 head/dimension/Top-K/length configurations, and alternated measurement order. Most tested n128 cases reduced latency by 12–62%; n128 H64/D512/Top-K128 was essentially unchanged. Compact H128/D512/Top-K512 (also n128) improved by 19%. Tested n64 cases were approximately unchanged in this smaller sweep; the earlier Sq4096/Sk128K n64 provided case slowed from 521 to 572 us at O2, motivating the n128-only policy. Full outputs were bitwise equal between O2 and O3 across all 15 sweep cases.

Before/after PR benchmark output samples also passed comparison; fused Top-K samples compare complete selected rows after sorting by ID, allowing only small floating-point softmax differences (rtol=1e-5, atol=1e-6).

Diagnostic NCU evidence for the historical H64/D512 n128 case: default 4.6.2 used 168 registers and issued 2,169,428 local load requests / 442,344 local store requests; O2 used 130 registers with zero local load/store requests. O2 and O3 emitted identical PTX, and assembling that same PTX with one system ptxas reproduced spill at O3 and no spill at O2. This diagnosis is distinct from the PR wrapper timings above.

H100/GB200/GB300 were not rerun for this addition. The Flex Attention GB300 figures remain **pre-migration reference data**, not measurements of this PR. Full DSA sweeps were not run.

### Formatting and documentation checks

- Black 26.3.1 / Black Jupyter (line length 160), SPDX headers, and `git diff --check` passed for the changed files. No C++ formatting applies to this addition.
- All 11 relative links in the Flex Attention API/design documents resolve; the three PNGs match source checkout `9c6cbf7` byte-for-byte (earlier validation in this PR).

## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] Formatting and applicable checks passed.
- [x] Reviewed the applicable root, Python, and test `AGENTS.md` Hard Rules.
- [x] Labels include `cat-enhancements`, `cat-doc`, `area:sparse_attention`, `op: DSA`, and `orig-nv-eng`.

## Summary by CodeRabbit

- **Documentation**
  - Added comprehensive Flex Attention mask-planning design documentation, benchmark results, and performance figures.
  - Linked Flex Attention planning documentation from the API overview and documentation index.
  - Expanded DSA guidance for local/global IDs, deterministic sparse backward behavior, Top-K configuration, and microbatching.
  - Clarified indexer behavior for local ID validation and conversion.

- **Bug Fixes**
  - Improved SM100 sparse-attention handling for local Top-K IDs and score recomputation across supported configurations.

- **Tests**
  - Added regression coverage for SM100 local-ID gathering and score-recomputation correctness.

### CODE REVIEW

**Comment by jiayus-nvidia** [comment]
@cudnn-ci-bot run oss

**Comment by jiayus-nvidia** [comment]
@cudnn-ci-bot run oss

**Review by Anerudhan — APPROVED** [review]
